luigi_pipeline/lib/model/seqr_mt_schema.py

```diff
@@ -0,0 +1,117 @@
+
+import hail as hl
+
+from lib.model.base_mt_schema import BaseMTSchema, mt_annotation
+from hail_scripts.v02.utils.computed_fields import variant_id
+from hail_scripts.v02.utils.computed_fields import vep
+
+
+class SeqrSchema(BaseMTSchema):
+
+    @mt_annotation(annotation='sortedTranscriptConsequences')
+    def sorted_transcript_consequences(self):
+        return vep.get_expr_for_vep_sorted_transcript_consequences_array(self.mt.vep)
+
+    @mt_annotation(annotation='docId', fn_require='sorted_transcript_consequences')
+    def doc_id(self, length=512):
+        return variant_id.get_expr_for_variant_id(self.mt, length)
+
+    @mt_annotation(annotation='variantId', fn_require='sorted_transcript_consequences')
+    def variant_id(self):
+        return variant_id.get_expr_for_variant_id(self.mt)
+
+    @mt_annotation(fn_require='sorted_transcript_consequences')
+    def contig(self):
+        return variant_id.get_expr_for_contig(self.mt.locus)
+
+    @mt_annotation(fn_require='sorted_transcript_consequences')
+    def pos(self):
+        return variant_id.get_expr_for_start_pos(self.mt)
+
+    @mt_annotation(fn_require='sorted_transcript_consequences')
+    def start(self):
+        return variant_id.get_expr_for_start_pos(self.mt)
+
+    @mt_annotation(fn_require='sorted_transcript_consequences')
+    def end(self):
+        return variant_id.get_expr_for_end_pos(self.mt)
+
+    @mt_annotation(fn_require='sorted_transcript_consequences')
+    def ref(self):
+        return variant_id.get_expr_for_ref_allele(self.mt)
+
+    @mt_annotation(fn_require='sorted_transcript_consequences')
+    def alt(self):
+        return variant_id.get_expr_for_alt_allele(self.mt)
+
+    @mt_annotation(fn_require='sorted_transcript_consequences')
+    def xpos(self):
+        return variant_id.get_expr_for_xpos(self.mt.locus)
+
+    @mt_annotation(fn_require='sorted_transcript_consequences')
+    def xstart(self):
+        return variant_id.get_expr_for_xpos(self.mt.locus)
+
+    @mt_annotation(fn_require='sorted_transcript_consequences')
+    def xstop(self):
+        return variant_id.get_expr_for_xpos(self.mt.locus) + hl.len(variant_id.get_expr_for_ref_allele(self.mt))
+
+    @mt_annotation(fn_require='sorted_transcript_consequences')
+    def domains(self):
+        return vep.get_expr_for_vep_protein_domains_set_from_sorted(
+            self.mt.sortedTranscriptConsequences)
```

Comment:
Hmm, this is where I'd disagree (assuming you mean calling `domains(dependency=`sortedTranscriptConsequences )`), since the dependency should be a property of the annotation itself, not something that should be changed, and should be fixed in the schema. You shouldn't have 2 callers have different dependencies. 

If you mean changing it so it's `@mt_annotation(annotation_require='sortedTranscriptConsequences')`, I agree that would be better, but would be more complex since it require the explicit mapping between annotation and function. This wouldn't be impossible to do, but I think it would add unnecessary complexity for now since in practice, I think the function name should be the annotation name (ours is weird since I have to support legacy stuff).

Reply:
I was thinking something like
```python
@mt_annotation(fn_require='sorted_transcript_consequences')
def domains(self, sorted_transcript_consequences):
  return vep.get_expr_for_vep_protein_domains_set_from_sorted(sorted_transcript_consequences)
```

where the `sorted_transcript_consequences` passed to `domains` is `mt.sortedTranscriptConsequences`.

Thus, if the annotation name on `sorted_transcript_consequences` is changed to `foo`, domains doesn't have to be changed to reference `mt.foo`.

Reply:
Ahh interesting idea. And so the arg `sorted_transcript_consequences` would be the return value of that function? That would require some explicit memoization mapping, and the implication is the client can override this (not sure if this is a good or bad thing). It's definitely interesting and I can see it's elegance. Let me sleep on it and if we decide to go with that we'll do another PR.

Reply:
I don't think it would be difficult to implement. Store the return value of `getattr(self, fn_require.__name__)()` and pass it to `func` in `wrapper`.

> That would require some explicit memoization mapping, and the implication is the client can override this (not sure if this is a good or bad thing)
Yeah, I think the difficult part is how to handle arguments when calling individual annotation methods.

Reply:
Right, because the function expects the arg to be an mt root, but the caller would not have access to that, so we'd have to make it optional e.g. 

`def domains(self, sorted_transcript_consequences=None):`

However, one we actually need to pass the value, we'd have to know the name of the named argument? This also makes it weird because we want to support the ability for the caller to pass in its own vars to the function (e.g. config vars, or other mt's to join).